### ISSUE TITLE
combineReducers does not support root immutable state

### ISSUE THREAD

**Issue by leonaves** [raisemarketplace/redux-loop#38]
The implementation of combineReducers provided with this library does not support a root immutable state such as Map. (Which is odd as the example in the repo uses a Map as its root state—but doesn't use `combineReducers`). The problematic areas are here in `combineReducers.js`:

```javascript
// line 22
const previousStateForKey = state[key];
```

And the mapValues function in `utils.js`:

```javascript
export function mapValues(object, mapper) {
  return Object.keys(object).reduce((current, key) => {
    current[key] = mapper(object[key], key);
    return current;
  }, {});
}
```

This limits the root state to being a plain object. I have written my own version of these two functions for my project, but it's a bit awkward as I have to use isLoop from the library like so:

```javascript
import { isLoop } from '../../../node_modules/redux-loop/lib/loop';
```

Which obviously could break at any time as this isn't an exposed function. Ideally, support for this would be built in. I was thinking about opening a pull request, but I wasn't sure how you might want to go about this (if you wanted it at all). Obviously I could drop in my function and export it as "`combineReducersImmutable`" or something similar. Or I could adjust the original function to check for immutable root states and to act appropriately.

Both these add code for one specific library however, which I've never been a huge fan of (but perhaps I'm overthinking it—it is a fairly popular library). So the last thing I thought of was to adjust combineReducers so it can accept any root data structure, something like:

```javascript
export function combineReducers(reducerMap, initialState = {}, accessor, mutator) {
    // code...
}

// and then to use:
const appReducer = combineReducers(reducers, Map(), 'get', 'set');
```

I may be jumping the shark here however (the accessor and mutator function names as strings makes me cringe—perhaps there's a better way? I'm just spitballing with this last suggestion really).

Anyway, would be interested to hear your thoughts, and happy to submit a pull request for an approach you prefer.

**Comment by leonaves** [comment]
Just thinking a bit more, another option might be to either export Symbol(isEffect) or to use Symbol.for() instead, as this is the main reason the function can't be replicated without requiring isLoop. (isLoop would be trivial to recreate if it didn't need that instance of that symbol).

**Comment by lukewestby** [comment]
I don't want to export `Symbol(isEffect)`, it's 100% an implementation detail. Exposing `isLoop` is definitely an option, I can think of plenty of other reasons why someone might want to have this function available in addition to this one.

Using immutable objects with combineReducers has come up before internally and in discussions so it would be great to find a solution. Like you said, including Immutable as a dependency isn't ideal so I wonder if we can iterate a little on your idea of passing an accessor and a mutator.

What if, instead of passing strings, which ultimately couples the implementation to the Immutable API or something identical to it in functionality, we allow for the passing of functions:

```javascript
const appReducer = combineReducers(
  reducers,
  Map(),
  (child, key) => child.get(key),
  (child, key, value) => child.set(key, value)
);
```

Thoughts?

**Comment by leonaves** [comment]
Seems to work well, I'm still trying to square it with changes made in [#36](https://github.com/raisemarketplace/redux-loop/pull/36), but I think it will work well. I was worried it was a bit long winded, but it adds no extra complexity to the default case of using an object (with proper defaults for initialState, accessor, and mutator.

I will try to make a pull request tomorrow, I would like to just get your or @claudiuandrei's input on [this question](https://github.com/raisemarketplace/redux-loop/pull/36#issuecomment-201100854), as I am a little lost as to why a reduce on `Object.keys(reducerMap)` is bad practice?

**Comment by leonaves** [comment]
https://github.com/raisemarketplace/redux-loop/pull/40

**Comment by lukewestby** [comment]
Published 2.1.0!
